src/reconcile/rec_visibility.c

```diff
@@ -331,8 +333,28 @@ __rec_find_and_save_delete_hs_upd(WT_SESSION_IMPL *session, WTI_RECONCILE *r, WT
             if (!seen_committed)
                 break;
             WT_RET(__rec_delete_hs_upd_save(session, r, ins, rip, delete_upd, NULL));
+            visible_all_upd = NULL;
             break;
         }
+
+        /*
+         * Track the first self-contained value that is globally visible. If we free the updates
+         * before the tombstone, due to the tombstone being globally visible concurrently with the
+         * update chain processing, this might be allowed to access the freed updates further in the
+         * reconciliation code.
+         *
+         * If a table has garbage collection enabled, then trim updates as possible. We should check
+         * the logic here - it might be possible to do something more aggressive?
+         */
+        if (!F_ISSET(r, WT_REC_EVICT) && visible_all_upd == NULL && delete_upd->next != NULL &&
```

Comment:
I think we should check F_ISSET(r, WT_REC_CHECKPOINT)? I don't think we need to run this in salvage.

Reply:
Done.